### PR TITLE
[AzureMonitorexporter] fix operation name

### PR DESCRIPTION
**Pull request by TimothyMothra**
When using Asp.Net Core Routing rules, our Exporter is currently recording the full pattern as the Operation Name.
This means that requests to different Controllers or Actions will show identical Operation Names, as shown in this screenshot.

Screenshot 2024-01-25 115910

This can be reproduced in an MVC app configured like this:
```
app.MapControllerRoute(
    name: "default",
    pattern: "{controller=Home}/{action=Index}/{id?}");
```

My test app has a `HomeController` and `CustomerController`. Regardless which controller is invoked, The `url.path` shows the correct Controller/Action. The `http.route` shows the configured pattern.

- https://localhost:44311/Home/Index

  ```
  Activity.ActivitySourceName: Microsoft.AspNetCore
  Activity.DisplayName:        GET {controller=Home}/{action=Index}/{id?}
  Activity.Kind:               Server
  Activity.Tags:
      server.address: localhost
      server.port: 44311
      http.request.method: GET
      url.scheme: https
      url.path: /
      network.protocol.version: 2
      http.route: {controller=Home}/{action=Index}/{id?}
      http.response.status_code: 200
  ```

- https://localhost:44311/Customer/Index

  ```
  Activity.ActivitySourceName: Microsoft.AspNetCore
  Activity.DisplayName:        GET {controller=Home}/{action=Index}/{id?}
  Activity.Kind:               Server
  Activity.Tags:
      server.address: localhost
      server.port: 44311
      http.request.method: GET
      url.scheme: https
      url.path: /Customer/Index
      network.protocol.version: 2
      http.route: {controller=Home}/{action=Index}/{id?}
      http.response.status_code: 200
  ```

This is in part due to the data provided by the OpenTelemetry.Instrumentation.AspNetCore library, where the configured pattern  is used as `http.route`. This is a known issue in that library: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/1730

### Changes
- adjust `TraceHelper` to better handle cases where `http.route` contains a controller pattern.
- Add more test cases

With this change:
image

### CODE REVIEW

**Comment by vishweshbankwar** [comment]
@TimothyMothra Could you confirm if the output with this change matches with the output when exporter without this change is used with 1.6.0-beta.3 version of the instrumentation library?

**Comment by TimothyMothra** [comment]
> @TimothyMothra Could you confirm if the output with this change matches with the output when exporter without this change is used with 1.6.0-beta.3 version of the instrumentation library?

Good catch. I needed to make the same change to the `GetOperationName` (v1) method. 

image

**Comment by bunkrur** [comment]
Thank you for this!

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/CHANGELOG.md [hunk 1]
```diff
@@ -8,6 +8,9 @@
 
 ### Bugs Fixed
 
+* Fix issue that caused Request OperationName to display routing pattern instead of url path.
```
Comment (vishweshbankwar) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/CHANGELOG.md:11]:
```suggestion
* Fix issue that caused Request OperationName to display identical routing pattern for different paths.
```

**Review by vishweshbankwar — APPROVED** [review]

File: sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/tests/Azure.Monitor.OpenTelemetry.AspNetCore.Demo/Controllers/CustomerController.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#if !NETFRAMEWORK
```
[the remark this answers is not in the document]
Comment (rajkumar-rangaraj) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/tests/Azure.Monitor.OpenTelemetry.AspNetCore.Demo/Controllers/CustomerController.cs:4]: @TimothyMothra Please don't make the demo complex.

**Review by rajkumar-rangaraj — APPROVED** [review]

**Comment by insylogo** [comment]
Routes should not be fully populated if you don't want to blow up cardinality on requests i.e. when you go to app insights, you should see something like /Controller/Method/{id} and not /Controller/Method/123, otherwise you can't get accurate statistics in most views for request timing etc

File: sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/tests/Azure.Monitor.OpenTelemetry.AspNetCore.Demo/Controllers/CustomerController.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (TimothyMothra) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/tests/Azure.Monitor.OpenTelemetry.AspNetCore.Demo/Controllers/CustomerController.cs:4]: I made the changes while working so I could reproduce the issue. Will remove the changes now.

**Comment by TimothyMothra** [comment]
Closing this PR for now. Tracking the issue here: https://github.com/Azure/azure-sdk-for-net/issues/41650
Were going to work with the OpenTelemetry community to see what fixes can be applied there.

**Pull request closed without merging**